SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialPackageMapClient.cpp

```diff
@@ -519,14 +519,11 @@ FNetworkGUID FSpatialNetGUIDCache::AssignNewStablyNamedObjectNetGUID(UObject* Ob
 			   TEXT("Found object called PersistentLevel which isn't a Level! This is not allowed when using the GDK"));
 	}
 
-	bool bNoLoadOnClient = false;
-	if (IsNetGUIDAuthority())
-	{
-		// If the server is replicating references to things inside levels, it needs to indicate
-		// that the client should not load these. Once the level is streamed in, the client will
-		// resolve the references.
-		bNoLoadOnClient = !CanClientLoadObject(Object, NetGUID);
-	}
+	// It is important we set this value correctly regardless of if we are the client or the server.
+	// It might be that the client has streamed a sub-level that the server has not yet told it about.
+	// This means the client will be registering the ObjectRef itself and will not cache the servers values.
+	bool bNoLoadOnClient = !CanClientLoadObject(Object, NetGUID);
```

Comment:
I'm concerned about this change - the reason native unreal doesn't allow clients to call it is being discarded at this point. Looking into the function though, I'm not sure what aspect of it would fail/be inconsistent if called from a client though, however @improbable-valentyn might have a better idea. And if we are to go with this approach we should remove the additional bool we store in the UnrealObjRef.

Reply:
I agree that `CanClientLoadObject` seems like it can be called from the clients. It seems that the reason we get a discrepancy is because we always go through `UObject*` -> `NetGUID` -> `UnrealObjectRef` chain and allocate NetGUIDs for stably named objects when calling RPCs from the client, whereas a cleaner approach would perhaps be to have a direct conversion from `UObject*` to `UnrealObjectRef` with the object path that doesn't involve creating the cache entry. That said, I think we shouldn't go for that now since that would be a riskier change, while it seems like the current solution is unlikely to introduce new issues.
Regarding your last point: we can't remove the bool from `UnrealObjectRef` because that's necessary for when the client receives the reference to something in the level from the server for the first time. At that time, the object still isn't loaded, so the client can't determine if it can load it itself.